### PR TITLE
Add SaLunamedia bidder

### PR DESCRIPTION
**Pull request by And1sS**

### CODE REVIEW

File: src/main/java/org/prebid/server/bidder/salunamedia/SaLunamediaBidder.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        final SeatBid firstSeatBid = seatBids.get(0);
+        final List<Bid> bids = firstSeatBid != null ? firstSeatBid.getBid() : null;
+        if (CollectionUtils.isEmpty(bids)) {
+            throw new PreBidException("Empty SeatBid.Bids");
+        }
+
+        final Bid firstBid = bids.get(0);
+        final ObjectNode firstBidExt = firstBid != null ? firstBid.getExt() : null;
+        final BidType firstBidType = firstBidExt != null ? getBidType(firstBidExt) : null;
+        if (firstBidType == null) {
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/salunamedia/SaLunamediaBidder.java:75]: It is possible case when we have ext, but not valid bidType, then we will have message `Missing BidExt`. Pls update this logic

File: src/main/java/org/prebid/server/bidder/salunamedia/SaLunamediaBidder.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+            throw new PreBidException("Missing BidExt");
+        }
+        return Collections.singletonList(BidderBid.of(firstBid, getBidType(firstBidExt), bidResponse.getCur()));
+    }
+
+    private BidType getBidType(ObjectNode node) {
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/salunamedia/SaLunamediaBidder.java:81]: Pls rename `node` to `bidExt`

File: src/test/java/org/prebid/server/bidder/salunamedia/SaLunamediaBidderTest.java [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        // when
+        final Result<List<BidderBid>> result = saLunamediaBidder.makeBids(httpCall, null);
+
+        // then
+        assertThat(result.getValue()).isEmpty();
+        assertThat(result.getErrors()).containsExactlyInAnyOrder(BidderError.badServerResponse("Missing BidExt"));
+    }
+
+    @Test
+    public void makeBidsShouldReturnOnlyFirstBid() throws JsonProcessingException {
```
Comment (snahornyi) [inline src/test/java/org/prebid/server/bidder/salunamedia/SaLunamediaBidderTest.java:165]: makeBidsShouldReturnOnlyBidderBidWithFirstBid

File: src/test/java/org/prebid/server/bidder/salunamedia/SaLunamediaBidderTest.java [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+                        bidBuilder -> bidBuilder.impid("123").ext(mediaTypeObjectNode),
+                        bidBuilder -> bidBuilder.impid("345").ext(mediaTypeObjectNode)))));
+
+        // when
+        final Result<List<BidderBid>> result = saLunamediaBidder.makeBids(httpCall, null);
+
+        // then
+        assertThat(result.getErrors()).isEmpty();
+        assertThat(result.getValue()).containsExactly(BidderBid.of(
+                Bid.builder().impid("123").ext(mediaTypeObjectNode).build(), xNative, null));
```
Comment (snahornyi) [inline src/test/java/org/prebid/server/bidder/salunamedia/SaLunamediaBidderTest.java:182]: Pls add for banner and video types too

File: src/test/java/org/prebid/server/bidder/salunamedia/SaLunamediaBidderTest.java [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+    private static BidRequest givenBidRequest(
+            Function<BidRequest.BidRequestBuilder, BidRequest.BidRequestBuilder> bidRequestCustomizer,
+            Function<Imp.ImpBuilder, Imp.ImpBuilder> impCustomizer) {
+
+        return bidRequestCustomizer.apply(BidRequest.builder()
+                .imp(singletonList(givenImp(impCustomizer))))
+                .build();
+    }
+
+    private static BidRequest givenBidRequest(Function<Imp.ImpBuilder, Imp.ImpBuilder> impCustomizer) {
```
Comment (snahornyi) [inline src/test/java/org/prebid/server/bidder/salunamedia/SaLunamediaBidderTest.java:194]: This method is never used

File: src/test/java/org/prebid/server/bidder/salunamedia/SaLunamediaBidderTest.java [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+                        .build(),
+                mapper.writeValueAsString(givenBidResponses(singletonList(bidBuilder -> bidBuilder.impid("123").ext(
+                        jacksonMapper.mapper().createObjectNode().put("mediaType", "invalid_type "))))));
+
+        // when
+        final Result<List<BidderBid>> result = saLunamediaBidder.makeBids(httpCall, null);
+
+        // then
+        assertThat(result.getValue()).isEmpty();
+        assertThat(result.getErrors()).containsExactlyInAnyOrder(BidderError.badServerResponse("Missing BidExt"));
```
Comment (snahornyi) [inline src/test/java/org/prebid/server/bidder/salunamedia/SaLunamediaBidderTest.java:161]: ContainsExactly is enough

**Review by snahornyi — CHANGES_REQUESTED** [review]

**Review by snahornyi — DISMISSED** [review]

File: src/test/java/org/prebid/server/bidder/salunamedia/SaLunamediaBidderTest.java [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+import static java.util.Collections.singletonList;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatIllegalArgumentException;
+import static org.prebid.server.proto.openrtb.ext.response.BidType.banner;
+import static org.prebid.server.proto.openrtb.ext.response.BidType.video;
+import static org.prebid.server.proto.openrtb.ext.response.BidType.xNative;
+
+public class SaLunamediaBidderTest extends VertxTest {
+
+    private static final String ENDPOINT_URL = "https://mobile.mng-ads.com/bidrequest{{ZoneID}}";
```
Comment (rpanchyk) [inline src/test/java/org/prebid/server/bidder/salunamedia/SaLunamediaBidderTest.java:36]: Minor. Pls don't use real urls in unit tests.

**Pull request merged**